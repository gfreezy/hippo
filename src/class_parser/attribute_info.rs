mod predefined_attribute;
use nom::eof;
use nom::multi::length_data;
use nom::number::complete::{be_u16, be_u32};
use nom::IResult;

#[derive(Debug)]
pub struct AttributeInfo {
    pub attribute_name_index: u16,
    pub info: Vec<u8>,
}

pub fn parse_attribute_info(buf: &[u8]) -> IResult<&[u8], AttributeInfo> {
    let (left, attribute_name_index) = be_u16(buf)?;
    let (left, info) = length_data(be_u32)(left)?;
    Ok((
        left,
        AttributeInfo {
            attribute_name_index,
            info: info.to_vec(),
        },
    ))
}

#[cfg(test)]
mod tests {
    use crate::class_parser::attribute_info::parse_attribute_info;
    use insta::assert_debug_snapshot;
    use nom::multi::many_m_n;

    #[test]
    fn test_parse_attribute_info() {
        let buf = [0, 13, 0, 0, 0, 2, 0, 14];
        let parser = many_m_n(1, 1, parse_attribute_info);
        assert_debug_snapshot!(parser(&buf).unwrap());
    }
}
