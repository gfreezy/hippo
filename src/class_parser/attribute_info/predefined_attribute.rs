use crate::class_parser::attribute_info::AttributeInfo;

pub struct ConstantValueAttribute {
    constant_value_index: u16,
}

pub struct ExceptionHandler {
    start_pc: u16,
    end_pc: u16,
    handler_pc: u16,
    catch_type: u16,
}

pub struct CodeAttribute {
    max_stack: u16,
    max_locals: u16,
    code: Vec<u8>,
    exception_table: Vec<ExceptionHandler>,
    attributes: Vec<AttributeInfo>,
}
