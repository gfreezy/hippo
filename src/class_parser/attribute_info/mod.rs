#[macro_use]
pub mod predefined_attribute;

use crate::class_parser::attribute_info::predefined_attribute::{
    parse_predefined_attribute, PredefinedAttribute,
};
use crate::class_parser::constant_pool::ConstPool;
use nom::number::complete::{be_u16, be_u32};
use nom::IResult;

#[derive(Debug)]
pub struct AttributeInfo {
    pub attribute_name_index: u16,
    pub attribute: PredefinedAttribute,
}

pub fn parse_attribute_info<'a>(
    const_pool: &ConstPool,
    buf: &'a [u8],
) -> IResult<&'a [u8], AttributeInfo> {
    let (buf, attribute_name_index) = be_u16(buf)?;
    let (buf, _length) = be_u32(buf)?;
    let attr_name = const_pool.get_utf8_string_at(attribute_name_index);
    let (buf, attr) = parse_predefined_attribute(attr_name, const_pool, buf)?;
    Ok((
        buf,
        AttributeInfo {
            attribute_name_index,
            attribute: attr,
        },
    ))
}
