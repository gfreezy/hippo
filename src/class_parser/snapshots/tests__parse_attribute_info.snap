---
source: src/class_parser/attribute_info.rs
expression: parser(&buf).unwrap()
---
(
    [],
    [
        AttributeInfo {
            attribute_name_index: 13,
            info: [
                0,
                14,
            ],
        },
    ],
)
